x-pack/filebeat/input/streaming/websocket.go

```diff
@@ -58,6 +65,47 @@ func (l *loggingRoundTripper) RoundTrip(req *http.Request) (*http.Response, erro
 	return resp, err
 }
 
+// heartBeat sends a ping message to the websocket server at regular intervals
+func (k *keepAlive) heartBeat(ctx context.Context, conn *websocket.Conn, start time.Time) context.CancelFunc {
+	ctx, cancel := context.WithCancel(ctx)
+	// set initial read deadline
+	conn.SetReadDeadline(time.Now().Add(k.cfg.KeepAlive.readControlDeadline))
+	// set pong handler to update read deadline
+	conn.SetPongHandler(func(string) error {
+		k.log.Debugw("received pong message from websocket server")
+		k.metrics.pongMessageReceivedTime.Update(time.Since(start).Nanoseconds())
+		return conn.SetReadDeadline(time.Now().Add(k.cfg.KeepAlive.readControlDeadline))
+	})
+
+	// set heartbeat ping routine
+	go func() {
+		ticker := time.NewTicker(k.cfg.KeepAlive.Interval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				if conn != nil {
+					err := conn.WriteControl(websocket.PingMessage, []byte{}, time.Now().Add(k.cfg.KeepAlive.WriteControlDeadline))
+					if err != nil {
+						k.log.Debugw("error sending ping message to websocket server:", err)
+						return
```

Comment:
Do we want to return here? Is there anything else that we need to do other than just taking our bat and ball and going home?

Reply:
No this `return` should be removed. We wan't the heartbeat routine to continue irrespective of write_control errors, and let our retry logic decide when to. quit. If such a scenario occurs when the infinite & blanket retries are also simultaneously enabled - (this is clearly not recommended in the doc), these debug logs will indicate the server is clearly unresponsive.